apps/remix-ide/src/app/files/fileManager.js

```diff
@@ -213,21 +213,58 @@ class FileManager extends Plugin {
    * @param {string} dest path of the destrination file
    * @returns {void}
    */
-  async copyFile (src, dest) {
+  async copyFile (src, dest, customName) {
     try {
       src = this.limitPluginScope(src)
       dest = this.limitPluginScope(dest)
-      await this._handleExists(src, `Cannot copy from ${src}`)
-      await this._handleIsFile(src, `Cannot copy from ${src}`)
-      await this._handleIsFile(dest, `Cannot paste content into ${dest}`)
+      await this._handleExists(src, `Cannot copy from ${src}. Path does not exist.`)
+      await this._handleIsFile(src, `Cannot copy from ${src}. Path is not a file.`)
+      await this._handleExists(dest, `Cannot paste content into ${dest}. Path does not exist.`)
+      await this._handleIsDir(dest, `Cannot paste content into ${dest}. Path is not directory.`)
       const content = await this.readFile(src)
+      const copiedFileName = customName ? '/' + customName : '/' + `Copy_${helper.extractNameFromKey(src)}`
 
-      await this.writeFile(dest, content)
+      await this.writeFile(dest + copiedFileName, content)
     } catch (e) {
       throw new Error(e)
     }
   }
 
+  /**
+   * Upsert a directory with the content of the source directory
+   * @param {string} src path of the source dir
+   * @param {string} dest path of the destination dir
+   * @returns {void}
+   */
+  async copyDir (src, dest) {
+    try {
+      src = this.limitPluginScope(src)
+      dest = this.limitPluginScope(dest)
+      await this._handleExists(src, `Cannot copy from ${src}. Path does not exist.`)
+      await this._handleIsDir(src, `Cannot copy from ${src}. Path is not a directory.`)
+      await this._handleExists(dest, `Cannot paste content into ${dest}. Path does not exist.`)
+      await this._handleIsDir(dest, `Cannot paste content into ${dest}. Path is not directory.`)
+      await this.inDepthCopy(src, dest)
+    } catch (e) {
+      throw new Error(e)
+    }
+  }
+
+  async inDepthCopy (src, dest, count = 0) {
```

Comment:
it might also be good to at some point to use https://github.com/ethereum/remix-project/blob/master/apps/remix-ide/src/app/files/fileProvider.js#L242 `copyFolderToJson`